### PR TITLE
RDCC-5417 - Controller, service, serviceImpl, JunitTest cases, Functional Test Case, PACT Test Cases, Sonar Violation fixes

### PR DESCRIPTION
**Pull request by vilasshelke-hmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDCC-5417

### Change description ###

Staff : Create New API to retrieve the user type for users

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

**Review by shivajihmcts — APPROVED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    @SuppressWarnings("unchecked")
+    @Test
+    @ToggleEnable(mapKey = mapKey, withFeature = true)
+    void fetchUserTypeAllStatusCode_200() {
+        StaffRefDataUserTypesResponse responses = (StaffRefDataUserTypesResponse)
+                caseWorkerApiClient.fetchUserType(
+                HttpStatus.OK
+            );
+        assertTrue(responses.getUserTypes().size() > 0);
```
Comment (rajkatla-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:74]: 1. can you add assertions for expected number of user types and their content.
2. add another test case for empty user types

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/client/CaseWorkerApiClient.java [hunk 2]
```diff
@@ -217,4 +220,19 @@ public Response deleteCaseworkerByIdOrEmailPattern(String path, HttpStatus statu
...
+        Response response = getMultipleAuthHeadersInternal(ROLE_CWD_ADMIN)
+                .get("/refdata/case-worker/user-type")
+                .andReturn();
+        response.then()
+                .assertThat()
+                .statusCode(expectedStatus.value());
+        if (expectedStatus.is2xxSuccessful()) {
+            return response.getBody().as(StaffRefDataUserTypesResponse.class);
+        } else {
+            return response.getBody().as(ErrorResponse.class);
```
Comment (rajkatla-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/client/CaseWorkerApiClient.java:234]: may i know in what scenario the else block invokes ?

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 3]
```diff
@@ -19,4 +39,61 @@ public class StaffRefDataController {
...
+        if (ObjectUtils.isNotEmpty(userTypes)) {
+            List<StaffRefDataUserType> refDataUserTypes = userTypes.stream()
+                    .map(StaffRefDataUserType::new)
+                    .toList();
+            staffReferenceDataUserTypesResponseBuilder.userTypes(refDataUserTypes);
+            log.debug("refDataUserTypes = {}", refDataUserTypes);
+            return ResponseEntity
+                    .status(200)
+                    .body(staffReferenceDataUserTypesResponseBuilder.build());
+        } else {
```
Comment (rajkatla-hmcts) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java:94]: do we need to return 404 ? I thought endpoint should return 200 for empty user types as well. check with Sujeeth on this.

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        responseEntity = ResponseEntity.status(200).body(null);
+        when(staffRefDataServiceMock.fetchUserTypes())
+                .thenReturn(userTypes);
+
+        ResponseEntity<?> actual = staffRefDataController.fetchUserTypes();
+
+        assertNotNull(actual);
+        verify(staffRefDataServiceMock, times(1))
+                .fetchUserTypes();
+        assertEquals(responseEntity.getStatusCode(), actual.getStatusCode());
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:71]: can check for the expected user types count and their content as well

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    }
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:88]: check for the count and empty array

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    private StaffRefDataServiceImpl staffRefDataService;
+
+    @Test
+    @SuppressWarnings("unchecked")
+    void testFetchUserType_All() {
+        when(userTypeRepository.findAll())
+                .thenReturn(prepareUserTypeResponse());
+        var userTypes = staffRefDataService
+                .fetchUserTypes();
+        verifyAllUserTypes(userTypes);
```
Comment (rajkatla-hmcts) [inline src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceTest.java:35]: assert the expected count and their all content

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 7]
```diff
@@ -19,4 +39,61 @@ public class StaffRefDataController {
...
+                    code = 401,
+                    message = UNAUTHORIZED_ERROR
+            ),
+            @ApiResponse(
+                    code = 403,
+                    message = FORBIDDEN_ERROR
+            ),
+            @ApiResponse(
+                    code = 404,
+                    message = "The User Types could not be found"
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java:67]: Can you please remove this as it is not required?

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java (hunk 3 above)
Comment (SujeethKumarAri) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java:94]: yes, it should return 200 with empty response

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 8]
```diff
@@ -19,4 +39,61 @@ public class StaffRefDataController {
...
+            @ApiResponse(
+                    code = 500,
+                    message = INTERNAL_SERVER_ERROR
+            )
+    })
+    @GetMapping(
+            path = "/user-type",
+            produces = APPLICATION_JSON_VALUE
+    )
+    @Secured("cwd-admin")
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java:78]: Please secure with "staff-admin"

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java (hunk 1 above)
Comment (vilasshelke-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:74]: @rajkatla-hmcts - regarding point 1 - don't you think this will be tightly couple with the data? Let say if data will be add / delete / update this case will be fail. Then we need to rewrite the test case. Is this correct approach?

Comment (rajkatla-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:74]: @vilasshelke-hmcts ok, ignore first comment

Comment (vilasshelke-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:74]: point 2 - This case I have handled in JUNIT (StaffRefDataControllerTest.shouldFetchNoUserTypes).

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java (hunk 5 above)
Comment (vilasshelke-hmcts) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:110]: @rajkatla-hmcts - Added

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceTest.java (hunk 6 above)
Comment (vilasshelke-hmcts) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/service/impl/StaffRefDataServiceTest.java:35]: @rajkatla-hmcts - Added

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java [hunk 9]
```diff
@@ -19,4 +38,61 @@ public class StaffRefDataController {
...
+                    code = 401,
+                    message = UNAUTHORIZED_ERROR
+            ),
+            @ApiResponse(
+                    code = 403,
+                    message = FORBIDDEN_ERROR
+            ),
+            @ApiResponse(
+                    code = 404,
+                    message = "The User Types could not be found"
```
[the remark this answers is not in the document]
Comment (vilasshelke-hmcts) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataController.java]: @SujeethKumarAri  - removed

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java (hunk 5 above)
Comment (vilasshelke-hmcts) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/StaffRefDataControllerTest.java:110]: @rajkatla-hmcts - this was added due to 404 with failure but now removed else block

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                );
+        assertTrue(responses.getUserTypes().size() > 0);
+    }
+
+    @Test
+    @ExtendWith(FeatureToggleConditionExtension.class)
+    @ToggleEnable(mapKey = mapKey, withFeature = false)
+    void should_fetchUserTypes_403_when_Api_toggled_off() {
+
+        Response response = caseWorkerApiClient.getMultipleAuthHeadersInternal(ROLE_CWD_SYSTEM_USER)
```
Comment (SujeethKumarAri) [inline src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:82]: You need to use "staff-admin" token

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java (hunk 10 above)
Comment (vilasshelke-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:82]: @SujeethKumarAri - this test case is for 403.

Comment (SujeethKumarAri) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:82]: yes, you need to use the same token (staff-admin) as 200 scenario. The only difference is that you have toggled off the api.
If you use the token other than staff-admin then you will always get 403 whether you have toggled on/off.

**Review by rajkatla-hmcts — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    void should_fetchUserTypes_403_when_Api_toggled_on() {
+
+        Response response = caseWorkerApiClient.getMultipleAuthHeadersInternal(ROLE_CWD_SYSTEM_USER)
+                .get("/refdata/case-worker/user-type")
+                .andReturn();
+        assertThat(HttpStatus.FORBIDDEN.value()).isEqualTo(response.statusCode());
+    }
+
+
+}
```
[the remark this answers is not in the document]
Comment (vilasshelke-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:101]: @rajkatla-hmcts  - point 1 is already covered in test case should_fetchUserTypes_403_when_Api_toggled_on

[the remark this answers is not in the document]
Comment (vilasshelke-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java:101]: @rajkatla-hmcts - We are not doing integration with any external micro-services. So I don't write integration test case.

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataUserTypesResponseTest.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        List<StaffRefDataUserType> refDataUserTypes = new ArrayList<>();
+        StaffRefDataUserType staffRefDataUserType = new StaffRefDataUserType();
+        staffRefDataUserType.setId(1L);
+        staffRefDataUserType.setCode("Test Code");
+        refDataUserTypes.add(staffRefDataUserType);
+        staffRefDataUserTypesResponse.setUserTypes(refDataUserTypes);
+        assertThat(staffRefDataUserTypesResponse.getUserTypes().get(0).getId()).isEqualTo(1L);
+        assertThat(staffRefDataUserTypesResponse.getUserTypes().get(0).getCode()).isEqualTo("Test Code");
+    }
+}
```
[the remark this answers is not in the document]
Comment (vilasshelke-hmcts) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataUserTypesResponseTest.java:26]: @rajkatla-hmcts - I think that we already covered in the PACT test

File: src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java (hunk 10 above)
Comment (vilasshelke-hmcts) [reply src/functionalTest/java/uk/gov/hmcts/reform/cwrdapi/StaffRefUserTypeFunctionalTest.java]: @SujeethKumarAri - updated

File: src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataUserTypesResponseTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (vilasshelke-hmcts) [reply src/test/java/uk/gov/hmcts/reform/cwrdapi/controllers/response/StaffRefDataUserTypesResponseTest.java:26]: @rajkatla-hmcts - this test case is belongs to response class. 200 code we already handled in controller.

**Review by rajkatla-hmcts — APPROVED** [review]

**Review by SujeethKumarAri — APPROVED** [review]
